Infrastructure for fine-grained debug flags (#13202)
coq/coq: doc/sphinx/proof-engine/vernacular-commands.rst

   interpreted from left to right, so in case of an overlap, the flags on the
   right have higher priority, meaning that `A,-A` is equivalent to `-A`.

.. opt:: Debug "{+, {? - } @ident }"

>    This option configures the display of debug messages. Debug
>    messages are separated in named components which may be activated
>    by being added to the list of identifiers, and deactivated by
>    prefixing their name by ``-``.
> 
>    A list of components and their current state can be obtained
>    through ``Test Debug``.
> 
>    The component ``all`` is special and activates or deactivates all other components.
> 
>    The component ``backtrace`` controls printing of error backtraces.
> 

Reviewer: ```suggestion
   Configures the display of debug messages. Each :n:`@ident` enables debug messages
   for that component, while  :n:`-@ident` disables messages for the component.
   ``all`` activates or deactivates all other components.  ``backtrace`` controls printing of
   error backtraces.

   :cmd:`Test` `Debug` displays the list of components and their enabled/disabled state.
```

   interpreted from left to right, so in case of an overlap, the flags on the
   right have higher priority, meaning that `A,-A` is equivalent to `-A`.

> .. opt:: Debug "{+, @ident {? = num } }"

Reviewer 2: ```suggestion
.. opt:: Debug "{+, @ident {? = @natural } }"
```

> .. opt:: Debug "{+, @ident {? = @natural } }"

Reviewer: Also, it's strange for `Warnings` and `Debug` to take a string, which then has to parsed with custom code rather than parsing that with the usual, well-tested main parser through productions in an mlg file.  It would be an additional rule for `Set`.  Please give some thought to this.

Reviewer 2: It's a very common pattern though. A significant part of the string options (`Bullet Behavior`, `Default Goal Selector`, `Default Proof Mode`, `Default Proof Using`, `Diffs`, `Loose Hint Behavior` and of course `Warnings`) interpret their string argument. It's also the case for one attribute (`using`).

Reviewer: How common is it? Except for `Default Goal Selector`, `Default Proof Using` and `Warnings`, the options you list choose a string from a small list of alternatives, for which a parser is hardly needed.  The first 3 would benefit from using the parser more directly.

The doc for `Default Goal Selector` doesn't match the code.  See `GoaL_select.parse_goal_selector`.  `Warnings` has its own custom parser.  `Default Proof Using` calls the main parser internally.

There is already one custom `Set` entry:
| "Set" "Firstorder" "Solver" ltac_expr

And there is this, too:
| "Debug" [ "On" | "Off" ]

`Debug` could at least use the main parser internally.

Reviewer 3: One advantage of having a custom (simpler) parser is that it can also be used for command line arguments.

Reviewer: I don't see the advantage.  You can easily parse a string corresponding to any nonterminal.

   This option configures the display of debug messages. Each debug
   message has a named component and a verbosity. When the debug level
   of the component is set at or higher than the verbosity of the
   message, the message is printed.

   In most cases, you do not need to know the verbosity of a message:
   ``Set Debug "component".`` will set it to the level of the main
   messages in the component.

   A list of components and their current levels can be obtained

>    through ``Test Debug.``.

Reviewer 2: The output looks funny with the two dots.

```suggestion
   through ``Test Debug``.
```

>    through ``Test Debug``.

Reviewer 2: `Test` shows the "value" of an option. The value of an option is not necessarily the last input it has received, especially in the case of string options which can be interpreted. AFAIK, the value of this particular option is just the verbosity level associated to each component.

Note that furthermore, this new option has an appending semantics, much like the `Warnings` option (which was the only such option thus far).

For warnings, this gives the following result:

```coq
Set Warnings "+deprecated".
Test Warnings.
(* Current value of Warnings is "+deprecated" *)
Set Warnings "-notations".
Test Warnings.
(* Current value of Warnings is "+deprecated,-notations" *)
Set Warnings "-deprecated".
Test Warnings.
(* Current value of Warnings is "-notations,-deprecated" *)
```

>    through ``Test Debug``.

Reviewer 4: `Warnings` in some sense prints a minimal string to get the current state from the initial state, but Debug just prints everything. So in the initial state it prints
~~~
"all=0,backtrace=0,unification=0,tactic-unification=0,proof-using=0,native-compiler=0,ho-unification=0,congruence=0,RAKAM=0,Cbv=0"
~~~

>    through ``Test Debug``.

Reviewer: > @Zimmi48 `Test` shows the "value" of an option.

`Set` is documented as simply assigning the given value to the option.  There's no description of other ways of handling the given value.

> (Set command) the option is set to the specified value.

> @SkySkimmer `Warnings` in some sense prints a minimal string to get the current state from the initial state

That behavior is OK, but the documentation doesn't mention any special handling for `Warnings` or `Debug`.

Perhaps someone would update the doc to reflect this?

>    through ``Test Debug``.

Reviewer 2: > > @Zimmi48 `Test` shows the "value" of an option.
> 
> `Set` is documented as simply assigning the given value to the option.  There's no description of other ways of handling the given value.
> 
> > (Set command) the option is set to the specified value.
> 
> > @SkySkimmer `Warnings` in some sense prints a minimal string to get the current state from the initial state
> 
> That behavior is OK, but the documentation doesn't mention any special handling for `Warnings` or `Debug`.
> 
> Perhaps someone would update the doc to reflect this?

Agreed.

Previously, I've complained about the exceptional behavior of `Set Warnings` with respect to the other options. But the development team has concluded that this type of behavior is actually useful in some cases, and this PR is a demonstration of that.

Therefore, it should be properly documented that `Set` can have an apprend behavior for a few options and it seems to me that it's best to do this as part of this very PR.

   The component ``all`` is special and sets all other components to
   its level. Its default level is ``1``, where most components

>    default to ``2``, so if you really want messages from all

Reviewer 2: Why not make its default `1` as well then?

Reviewer 4: While we still have the legacy messages I think we don't want -debug to make everything start spamming

   components you need ``Set Debug "all=2"``. Note that some debug
   messages have not been assigned a component and instead are printed
   when ``all`` is nonzero.

   The component ``backtrace`` is special: instead of controlling

>    debug messages it controls printing of error backtraces.

Reviewer 4: changed
